lib/internal/streams/duplexify.js

```diff
@@ -0,0 +1,335 @@
+      this._readableState.ended = true;
+      this._readableState.endEmitted = true;
+    }
+
+    if (options?.writable === false) {
+      this._writableState.writable = false;
+      this._writableState.ending = true;
+      this._writableState.ended = true;
+      this._writableState.finished = true;
+    }
+  }
+}
+
+module.exports = function duplexify(body, name) {
+  if (isDuplexNodeStream(body)) {
+    return body;
+  } else if (isReadableNodeStream(body)) {
+    return _duplexify({ readable: body });
+  } else if (isWritableNodeStream(body)) {
+    return _duplexify({ writable: body });
+  } else if (isNodeStream(body)) {
+    return _duplexify({ writable: false, readable: false });
+  } else if (isReadableStream(body)) {
+    return _duplexify({ readable: Readable.fromWeb(body) });
+  } else if (isWritableStream(body)) {
+    return _duplexify({ writable: Writable.fromWeb(body) });
+  } else if (typeof body === 'function') {
+    const { value, write, final } = fromAsyncGen(body);
+
+    if (isIterable(value)) {
+      return from(Duplexify, value, {
+        // TODO (ronag): highWaterMark?
+        objectMode: true,
+        write,
+        final
+      });
+    }
+
+    const { then } = value;
+    if (typeof then === 'function') {
+      let d;
+
+      const promise = FunctionPrototypeCall(
+        then,
+        value,
+        (val) => {
+          if (val != null) {
+            throw new ERR_INVALID_RETURN_VALUE('nully', 'body', val);
+          }
+        },
+        (err) => {
+          destroyer(d, err);
+        }
+      );
+
+      return d = new Duplexify({
+        // TODO (ronag): highWaterMark?
+        objectMode: true,
+        readable: false,
+        write,
+        final(cb) {
+          final(() => promise.then(cb, cb));
+        }
+      });
+    }
+
+    throw new ERR_INVALID_RETURN_VALUE(
+      'Iterable, AsyncIterable or AsyncFunction', name, value);
+  } else if (isBlob(body)) {
+    return duplexify(body.arrayBuffer());
+  } else if (isIterable(body)) {
+    return from(Duplexify, body, {
+      // TODO (ronag): highWaterMark?
+      objectMode: true,
+      writable: false
+    });
+  } else if (
+    isReadableStream(body?.readable) &&
+    isWritableStream(body?.writable)
+  ) {
+    return Duplexify.fromWeb(body);
+  } else if (
+    typeof body?.writable === 'object' ||
+    typeof body?.readable === 'object'
+  ) {
+    const readable = body?.readable ?
+      isReadableNodeStream(body?.readable) ? body?.readable :
+        duplexify(body.readable, body.options ?? body.readableOptions) :
+      undefined;
+
+    const writable = body?.writable ?
+      isWritableNodeStream(body?.writable) ? body?.writable :
+        duplexify(body.writable, body.options ?? body.writableOptions) :
+      undefined;
+
+    return _duplexify({ readable, writable });
+  } else if (typeof body?.then === 'function') {
+    let d;
+
+    FunctionPrototypeCall(
+      body.then,
```

Comment:
You shouldn't access `then` more than once.
```suggestion
  } else {
    const then = body?.then;
    if (typeof then === 'function') {
    let d;

    FunctionPrototypeCall(
      then,
```

Reply:
Why?

Reply:
Some Promise implementations throw if `then` is accessed more than once.